File: TrappingRainWater.java
class Solution {
    public int trap(int[] height) {
        int[] leftH = new int[height.length];
        int[] rightH = new int[height.length];

        int maxleft = 0;
        for(int i = 0; i < height.length; i++){
            leftH[i] = maxleft;
            if(height[i] > maxleft){
                maxleft = height[i];
            }
        }

        int maxright = 0;
        for(int i = height.length-1; i >= 0; i--){
            rightH[i] = maxright;
            if(height[i] > maxright){
                maxright = height[i];
            }
        }

        int count = 0;
        for(int i = 0; i < height.length; i++){
            int minHeight = Math.min(leftH[i], rightH[i]);
            int val = minHeight - height[i];
            if(val > 0){
                count += val;
            }
        }

        return count;
    }
}

File: TrappingRainWater.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        trapped = 0
        prev = [0] * len(height)
        after = [0] * len(height)

        maxleft = 0
        for i in range(len(height)):
            prev[i] = maxleft
            maxleft = max(height[i], maxleft)

        maxright = 0
        for i in range(len(height)-1, -1, -1):
            after[i] = maxright
            maxright = max(height[i], maxright)
        
        print(prev)
        print(after)
        for i in range(len(height)):
            temp = (min(after[i],prev[i])-height[i])
            if temp > 0:
                trapped += temp 
        return trapped